### PR TITLE
New-GcbqTable and formatting fixes

### PR DESCRIPTION
**Pull request by ahandley**
New cmdlet and tests, as well as a bunch of formatting fixes to dataset code and changing a parameter name to be consistent with documentation.

One area that I would like comments on - testing the -Expiration flag.  
Is there a better way to do that other than "set a table to expire 10 seconds from instantiation and wait 20 seconds after the call returns to check"?

### CODE REVIEW

File: Google.PowerShell.IntegrationTests/BigQuery/BigQuery.GcbqTable.Tests.ps1 [hunk 1]
```diff
@@ -1,56 +1,117 @@
 ﻿. $PSScriptRoot\..\BigQuery\GcbqCmdlets.ps1
 $project, $zone, $oldActiveConfig, $configName = Set-GCloudConfig
 
+$testDataset = "pshell_testing"
+
 Describe "Get-GcbqTable" {
 
-    #TODO(ahandley): Include setup and teardown code in these tests as New- and Remove- are developed.
+    $d = New-GcbqDataset $testDataset
```
Comment (quoctruong) [inline Google.PowerShell.IntegrationTests/BigQuery/BigQuery.GcbqTable.Tests.ps1]: Instead of pshell_testing, you should add a random variable to prevent collision. For example:
```
$r = Get-Random;
$testDataset = "pshell_testing_$r";
$dataset = New-GcbqDataset $testDataset";
```

Comment (quoctruong) [inline Google.PowerShell.IntegrationTests/BigQuery/BigQuery.GcbqTable.Tests.ps1]: Also, we should use something meaningful for variable name. For example $dataset instead of $d.

File: Google.PowerShell.IntegrationTests/BigQuery/BigQuery.GcbqTable.Tests.ps1 [hunk 2]
```diff
@@ -1,56 +1,117 @@
...
-        { Get-GcbqTable -project $nonExistProject -Dataset "pshell_testing"} | Should Throw "404"
+        { Get-GcbqTable -project $nonExistProject -Dataset $testDataset} | Should Throw "404"
+    }
+
+    $d | Remove-GcbqDataset -Force
+}
+
+Describe "New-GcbqTable" {
+
+    $d = New-GcbqDataset $testDataset
```
Comment (quoctruong) [inline Google.PowerShell.IntegrationTests/BigQuery/BigQuery.GcbqTable.Tests.ps1]: Here, you should initialize:
```
$r = Get-Random;
$testDataset = "pshell_testing_$r";
$dataset = New-GcbqDataset $testDataset";
```

File: Google.PowerShell.IntegrationTests/BigQuery/BigQuery.GcbqTable.Tests.ps1 [hunk 3]
```diff
@@ -1,56 +1,117 @@
...
+
+    It "should take a name, description, and time to make a table."{
+        $a = New-GcbqTable -Dataset $testDataset "my_table_1" -Name "CSV" -Description `
```
Comment (quoctruong) [inline Google.PowerShell.IntegrationTests/BigQuery/BigQuery.GcbqTable.Tests.ps1]: Please change $a to something meaningful like $table.

File: Google.PowerShell.IntegrationTests/BigQuery/BigQuery.GcbqDataset.Tests.ps1 [hunk 4]
```diff
@@ -3,7 +3,7 @@ $project, $zone, $oldActiveConfig, $configName = Set-GCloudConfig
 
 Describe "Get-GcbqDataset" {
 
-    It "should list datasets" {
+    It "should list datasets." {
```
Comment (quoctruong) [inline Google.PowerShell.IntegrationTests/BigQuery/BigQuery.GcbqDataset.Tests.ps1]: I don't think you really need to add "." at the end of this. If you look at the Pester [repo](https://github.com/PowerShell/PowerShell/blob/master/docs/testing-guidelines/WritingPesterTests.md), they don't have "." at the end of It statement.

File: Google.PowerShell.IntegrationTests/BigQuery/BigQuery.GcbqTable.Tests.ps1 [hunk 5]
```diff
@@ -1,56 +1,117 @@
...
     It "should throw when the dataset is not found."{
         { Get-GcbqTable -Dataset $nonExistDataset } | Should Throw "404"
     }
 
     It "should throw when the project is not found."{
-        { Get-GcbqTable -project $nonExistProject -Dataset "pshell_testing"} | Should Throw "404"
+        { Get-GcbqTable -project $nonExistProject -Dataset $testDataset} | Should Throw "404"
+    }
+
+    $d | Remove-GcbqDataset -Force
```
Comment (quoctruong) [inline Google.PowerShell.IntegrationTests/BigQuery/BigQuery.GcbqTable.Tests.ps1]: This should be in an "AfterAll" Pester block.

File: Google.PowerShell.IntegrationTests/BigQuery/BigQuery.GcbqTable.Tests.ps1 [hunk 6]
```diff
@@ -1,56 +1,117 @@
...
+
+    It "should throw when the dataset is not found."{
+        { New-GcbqTable "my_table_1" -Dataset $nonExistDataset } | Should Throw "404"
+    }
+
+    It "should throw when the project is not found."{
+        { New-GcbqTable -Dataset $testDataset "my_table_1" -project $nonExistProject} | Should Throw "404"
     }
+
+    $d | Remove-GcbqDataset -Force
```
Comment (quoctruong) [inline Google.PowerShell.IntegrationTests/BigQuery/BigQuery.GcbqTable.Tests.ps1]: This should be in "AfterAll" Pester block.

File: Google.PowerShell/BigQuery/GcbqDataset.cs [hunk 7]
```diff
@@ -375,7 +375,7 @@ protected override void ProcessRecord()
             catch (GoogleApiException ex) when (ex.HttpStatusCode == HttpStatusCode.Conflict)
             {
                 WriteError(new ErrorRecord(ex,  
-                    $"A dataset with the name {DatasetId} already exists.",
+                    $"A dataset with the name {DatasetId} already exists in this project.",
```
Comment (quoctruong) [inline Google.PowerShell/BigQuery/GcbqDataset.cs]: Since you actually have the Project's name, I think a better error message would be:
`$"A dataset with the name '{DatasetId}' already exists in project '{Project}'."`

File: Google.PowerShell/BigQuery/GcbqTable.cs [hunk 8]
```diff
@@ -112,4 +112,145 @@ protected override void ProcessRecord()
...
+            // Add the new dataset to the project supplied.
+            try
+            {
+                Table response = request.Execute();
+                WriteObject(response);
+            }
+            catch (GoogleApiException ex) when (ex.HttpStatusCode == HttpStatusCode.Conflict)
+            {
+                WriteError(new ErrorRecord(ex,
+                    $"A table with the name {TableId} already exists in this dataset.",
```
Comment (quoctruong) [inline Google.PowerShell/BigQuery/GcbqTable.cs]: It would be great if you can add the name of the dataset here:
`$"A table with the name '{TableId}' already exists in the dataset '{Dataset}'."`

File: Google.PowerShell/BigQuery/GcbqTable.cs [hunk 9]
```diff
@@ -112,4 +112,145 @@ protected override void ProcessRecord()
...
+        public string Description { get; set; }
+
+        /// <summary>
+        /// <para type="description">
+        /// The lifetime of this table from the time of creation (in seconds).
+        /// </para>
+        /// </summary>
+        [Parameter(Mandatory = false, ParameterSetName = ParameterSetNames.ByValues)]
+        [ValidateRange(3600, (long.MaxValue / 1000))]
+        public long Expiration { get; set; }
```
Comment (quoctruong) [inline Google.PowerShell/BigQuery/GcbqTable.cs]: It seems you don't have a test for this parameter. You should add it if that's the case.

File: Google.PowerShell/BigQuery/GcbqTable.cs [hunk 10]
```diff
@@ -112,4 +112,145 @@ protected override void ProcessRecord()
...
+                    newTab = new Table();
+                    newTab.TableReference = new TableReference();
+                    newTab.TableReference.DatasetId = Dataset;
+                    newTab.TableReference.ProjectId = Project;
+                    newTab.TableReference.TableId = TableId;
+                    newTab.FriendlyName = Name;
+                    newTab.Description = Description;
+                    if (Expiration != 0)
+                    {
+                        double currentMillis = DateTime.Now.ToUniversalTime().Subtract(
```
Comment (quoctruong) [inline Google.PowerShell/BigQuery/GcbqTable.cs]: You can use this built-in [method](https://msdn.microsoft.com/en-us/library/system.datetimeoffset.tounixtimemilliseconds(v=vs.110).aspx) to get milliseconds since epoch.

File: Google.PowerShell/BigQuery/GcbqTable.cs (hunk 9 above)
Comment (ahandley) [inline Google.PowerShell/BigQuery/GcbqTable.cs]: That's what I was asking about above.  I was planning to make one, but I don't know if there's a more elegant way than creating a table with a somewhat short life and then just calling thread.sleep() and checking later.

File: Google.PowerShell.IntegrationTests/BigQuery/BigQuery.GcbqTable.Tests.ps1 (hunk 1 above)
Comment (ahandley) [inline Google.PowerShell.IntegrationTests/BigQuery/BigQuery.GcbqTable.Tests.ps1]: Got it!

File: Google.PowerShell.IntegrationTests/BigQuery/BigQuery.GcbqDataset.Tests.ps1 (hunk 4 above)
Comment (ahandley) [inline Google.PowerShell.IntegrationTests/BigQuery/BigQuery.GcbqDataset.Tests.ps1]: Alright, fixed that.

File: Google.PowerShell/BigQuery/GcbqTable.cs (hunk 10 above)
Comment (ahandley) [inline Google.PowerShell/BigQuery/GcbqTable.cs]: So I tried using that method:
```
 DateTimeOffset d = new DateTimeOffset(DateTime.Now);
 newTab.ExpirationTime = d.ToUnixTimeMilliseconds();
```
and the compiler is coming up with "DateTimeOffset doesnt have a method 'ToUnixTimeMilliseconds()"... 
I have System imported, and it's recognizing the data type, but just not finding the method.  Any idea what could be causing this?

File: Google.PowerShell/BigQuery/GcbqTable.cs (hunk 9 above)
Comment (ahandley) [inline Google.PowerShell/BigQuery/GcbqTable.cs]: That's what I was asking about in the comment at the top.  I didn't want to write bad code, but that's the best way I can think of at the moment.

Comment (quoctruong) [inline Google.PowerShell/BigQuery/GcbqTable.cs]: You can create the table and check that the table has correct expiration time. The expiration time should be present in the table object right?

Comment (ahandley) [inline Google.PowerShell/BigQuery/GcbqTable.cs]: Yes, it is.  I wasnt sure if that was a sufficient test or not.  I'll code that up then.

File: Google.PowerShell/BigQuery/GcbqTable.cs (hunk 10 above)
Comment (quoctruong) [inline Google.PowerShell/BigQuery/GcbqTable.cs]: I think it's because that method is only in .NET 4.6.  In that case, we can just go with your current code.

File: Google.PowerShell/BigQuery/GcbqTable.cs (hunk 9 above)
Comment (ILMTitan) [inline Google.PowerShell/BigQuery/GcbqTable.cs]: It is sufficient because we are not testing that BigQuery works correctly. That is someone else's job, We are testing that our API works correctly.

File: Google.PowerShell/BigQuery/GcbqDataset.cs [hunk 11]
```diff
@@ -303,7 +303,7 @@ private class ParameterSetNames
         /// </para>
         /// </summary>
         [Parameter(Mandatory = false, ValueFromPipeline = true, ParameterSetName = ParameterSetNames.ByObject)]
-        public Dataset Dataset { get; set; }
+        public Dataset InputObject { get; set; }
```
Comment (ILMTitan) [inline Google.PowerShell/BigQuery/GcbqDataset.cs]: I'm not sure why this needed to be renamed. "InputObject" is an acceptable generic name, but "Dataset" is a better more specific name. Perhaps an alias is useful?

File: Google.PowerShell/BigQuery/GcbqTable.cs [hunk 12]
```diff
@@ -112,4 +112,145 @@ protected override void ProcessRecord()
...
+        [Parameter(Mandatory = true, ValueFromPipeline = true, ParameterSetName = ParameterSetNames.ByObject)]
+        public Table InputObject { get; set; }
+
+        /// <summary>
+        /// <para type="description">
+        /// The project to put the table in. If not set via PowerShell parameter processing, will
+        /// default to the Cloud SDK's DefaultProject property.
+        /// </para>
+        /// </summary>
+        [Parameter(Mandatory = false)]
```
Comment (ILMTitan) [inline Google.PowerShell/BigQuery/GcbqTable.cs]: This is only relevant to the parameter set ByValues.

File: Google.PowerShell/BigQuery/GcbqTable.cs [hunk 13]
```diff
@@ -112,4 +112,145 @@ protected override void ProcessRecord()
...
+        [ConfigPropertyName(CloudSdkSettings.CommonProperties.Project)]
+        override public string Project { get; set; }
```
Comment (ILMTitan) [inline Google.PowerShell/BigQuery/GcbqTable.cs]: This should be `public override string Project { get; set; }`

File: Google.PowerShell/BigQuery/GcbqTable.cs [hunk 14]
```diff
@@ -112,4 +112,145 @@ protected override void ProcessRecord()
...
+
+        /// <summary>
+        /// <para type="description">
+        /// The Dataset that you would like to add to. This field takes strings, Datasets, and DatasetRefrences.
+        /// </para>
+        /// </summary>
+        [Parameter(Mandatory = true, ValueFromPipeline = true, ParameterSetName = ParameterSetNames.ByValues)]
+        [PropertyByTypeTransformation(TypeToTransform = typeof(Dataset), Property = nameof(Apis.Bigquery.v2.Data.Dataset.DatasetReference))]
+        [PropertyByTypeTransformation(TypeToTransform = typeof(DatasetReference), Property = nameof(DatasetReference.DatasetId))]
```
Comment (ILMTitan) [inline Google.PowerShell/BigQuery/GcbqTable.cs]: You probably need a third ParameterSet to take a Dataset object rather than a string, because the `DatasetReference` will specify the project for you.

File: Google.PowerShell/BigQuery/GcbqDataset.cs (hunk 11 above)
Comment (ahandley) [inline Google.PowerShell/BigQuery/GcbqDataset.cs]: For consistency.  I did this because whenever i have an object that should be passed in via pipeline instead of by parameter, that's what I've been calling the variable.  I can go in and add aliases, but this is also the convention with other GCP cmdlets and it should be fine as is.

File: Google.PowerShell/BigQuery/GcbqTable.cs (hunk 12 above)
Comment (ahandley) [inline Google.PowerShell/BigQuery/GcbqTable.cs]: Had that in there earlier, but it got lost in debugging.  Thanks.

File: Google.PowerShell/BigQuery/GcbqTable.cs (hunk 14 above)
Comment (ahandley) [inline Google.PowerShell/BigQuery/GcbqTable.cs]: Added this in.  Going to submit another CL tomorrow and I'm going to update all previous cmdlets to support this.  I added a section to my design doc and I'm just going to make this the standard for referencing down the project->dataset->table chain.

**Comment by ahandley** [comment]
Ptal.  Everything above should have been addressed.

File: Google.PowerShell/BigQuery/GcbqTable.cs [hunk 15]
```diff
@@ -112,4 +112,179 @@ protected override void ProcessRecord()
...
+                    new DateTime(1970, 1, 1, 0, 0, 0, DateTimeKind.Utc)
+                    ).TotalMilliseconds);
+                newTab.ExpirationTime = (Expiration * 1000) + currentMillis;
+                // Note: The code below is a more elegant solution, but is not supported by the current version
+                //newTab.ExpirationTime = new DateTimeOffset(DateTime.Now).ToUnixTimeMilliseconds();
+            }
+            return Service.Tables.Insert(newTab, Project, DatasetId);
+        }
+
+        public TableReference getTableRef(string p, string d, string t)
```
Comment (ILMTitan) [inline Google.PowerShell/BigQuery/GcbqTable.cs]: These names should be full words: projectId, datasetId, tableId. The local variable should also probably be tableReference, but the function is short enough that it is not as important as the parameter names.

File: Google.PowerShell/BigQuery/GcbqTable.cs [hunk 16]
```diff
@@ -112,4 +112,179 @@ protected override void ProcessRecord()
...
+        {
+            TableReference tr = new TableReference();
```
Comment (ILMTitan) [inline Google.PowerShell/BigQuery/GcbqTable.cs]: You can make this easier with C#'s object initialization syntax

    return new TableReference { DatasetId = datasetId, ProjectId = projectId, ...};

You may not even need this function doing that.

File: Google.PowerShell/BigQuery/GcbqTable.cs [hunk 17]
```diff
@@ -112,4 +112,179 @@ protected override void ProcessRecord()
...
+                WriteError(new ErrorRecord(ex,
+                    $"A table with the name {TableId} already exists in {Project}:{DatasetId}.",
+                    ErrorCategory.InvalidArgument,
+                    TableId));
+            }
+        }
+
+        public TablesResource.InsertRequest makeInsertReq()
+        {
+            Table newTab;
```
Comment (ILMTitan) [inline Google.PowerShell/BigQuery/GcbqTable.cs]: There is no need to split declaration and assignment here. The variable name should be the full word newTable.

File: Google.PowerShell/BigQuery/GcbqTable.cs (hunk 16 above)
Comment (ahandley) [inline Google.PowerShell/BigQuery/GcbqTable.cs]: Yep.  Just killed the function and inlined it.  I originally made the function when I had to repeat that logic twice, but then I made a function out of the calling logic.

File: Google.PowerShell/BigQuery/GcbqTable.cs (hunk 17 above)
Comment (ahandley) [inline Google.PowerShell/BigQuery/GcbqTable.cs]: Got it.

File: Google.PowerShell.IntegrationTests/BigQuery/BigQuery.GcbqTable.Tests.ps1 [hunk 18]
```diff
@@ -3,53 +3,159 @@ $project, $zone, $oldActiveConfig, $configName = Set-GCloudConfig
...
         { Get-GcbqTable -Dataset $nonExistDataset } | Should Throw "404"
     }
 
-    It "should throw when the project is not found."{
-        { Get-GcbqTable -project $nonExistProject -Dataset "pshell_testing"} | Should Throw "404"
+    It "should throw when the project is not found"{
+        { Get-GcbqTable -project $nonExistProject -Dataset $datasetName} | Should Throw "404"
+    }
+
+    AfterAll{
```
Comment (ILMTitan) [inline Google.PowerShell.IntegrationTests/BigQuery/BigQuery.GcbqTable.Tests.ps1]: Nit: Space between `AfterAll` and `{`.

File: Google.PowerShell.IntegrationTests/BigQuery/BigQuery.GcbqTable.Tests.ps1 [hunk 19]
```diff
@@ -3,53 +3,159 @@ $project, $zone, $oldActiveConfig, $configName = Set-GCloudConfig
...
+
+    It "should throw when the dataset is not found"{
+        { New-GcbqTable "my_table_d404" -DatasetId $nonExistDataset } | Should Throw "404"
+    }
+
+    It "should throw when the project is not found"{
+        { New-GcbqTable -DatasetId $datasetName "my_table_p404" -project $nonExistProject} | Should Throw "404"
+    }
+
+    Afterall {
```
Comment (ILMTitan) [inline Google.PowerShell.IntegrationTests/BigQuery/BigQuery.GcbqTable.Tests.ps1]: Nit: `AfterAll` pascal case.

File: Google.PowerShell/BigQuery/GcbqTable.cs [hunk 20]
```diff
@@ -112,4 +112,174 @@ protected override void ProcessRecord()
...
+    /// <para type="description">
+    /// Creates a new empty table in the specified dataset. A Table can be supplied by object 
+    /// via the pipeline or the -InputObject parameter, or it can be instantiated by value 
+    /// with the flags below. The Dataset ID can be specified by passing in a string to 
+    /// -DatasetId, or you can pass a Dataset or DatasetReference to the -Dataset parameter. 
+    /// If no Project is specified, the default project will be used. This cmdlet returns 
+    /// a Table object.
+    /// </para>
+    /// <example>
+    ///   <code>PS C:\> New-GcbqTable “new_tab” -Dataset “my_data” -Description “Some nice data!”
```
Comment (ILMTitan) [inline Google.PowerShell/BigQuery/GcbqTable.cs]: Nit: Code blocks preserve whitespace.  Multi-line code examples should start on a new line so the intended alignment matches the actual alignment. 

This line should also have a back-tick ( \` ) character at the end to indicate it continues on the next line. Otherwise, someone who copies and pastes this example will get errors about `-Expiration` not being a valid operator.

File: Google.PowerShell/BigQuery/GcbqTable.cs [hunk 21]
```diff
@@ -112,4 +112,174 @@ protected override void ProcessRecord()
...
+        /// </para>
+        /// </summary>
+        [Parameter(Mandatory = false, ParameterSetName = ParameterSetNames.ByValues)]
+        [ConfigPropertyName(CloudSdkSettings.CommonProperties.Project)]
+        public override string Project { get; set; }
+
+        /// <summary>
+        /// <para type="description">
+        /// The DatasetId that you would like to add to. This field takes strings.
+        /// To pass in a Dataset or DatasetId object for this field, use the ByValuesWithRef parameter set.
```
Comment (ILMTitan) [inline Google.PowerShell/BigQuery/GcbqTable.cs]: I don't know what a DatasetId object is. Just say Dataset object. You don't need to talk about `DatasetReference` objects here.

Also, don't talk about the name of the parameter set. The name is not visible to the users. Instead, just mention the Dataset parameter.

**Comment by ahandley** [comment]
Thanks!
